sdk/opa.go

```diff
@@ -0,0 +1,405 @@
+	ready                  bool
+	mtx                    sync.Mutex
+}
+
+// ConfigFile sets the configuration file to use on the OPA instance.
+func ConfigFile(fileName string) func(*OPA) error {
+	return func(opa *OPA) error {
+		bs, err := ioutil.ReadFile(fileName)
+		if err != nil {
+			return err
+		}
+		opa.configBytes = bs
+		return nil
+	}
+}
+
+// Config sets the configuration to use on the OPA instance.
+func Config(config []byte) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.configBytes = config
+		return nil
+	}
+}
+
+// Logger sets the logging implementation to use for debug logs. Use this
+// function to configure a logger that is useful for debugging management API
+// errors:
+//
+//	o, err := opa.New(opa.Logger(logging.NewStandardLogger()))
+//	if err != nil { // handle error }
+func Logger(logger logging.Logger) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.logger = logger
+		return nil
+	}
+}
+
+// Console sets the logging implementation to use for console logs (e.g.,
+// status, decision logs, etc.)
+func Console(logger logging.Logger) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.console = logger
+		return nil
+	}
+}
+
+// Async causes opa.New() to return immediately instead of waiting for plugin
+// readiness. OPA will close the provided channel to indicate readiness. Use
+// this function to avoid blocking the goroutine that creates the OPA object.
+func Async(ch chan struct{}) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.async = true
+		opa.readyCh = ch
+		return nil
+	}
+}
+
+// New returns a new OPA object. This function should minimally be called with
+// an OPA configuration.
+func New(ctx context.Context, opts ...func(*OPA) error) (*OPA, error) {
+
+	opa := &OPA{
+		store:           inmem.New(),
+		preparedQueries: map[string]*rego.PreparedEvalQuery{},
+	}
+
+	for _, opt := range opts {
+		if err := opt(opa); err != nil {
+			return nil, err
+		}
+	}
+
+	if !opa.async {
+		opa.readyCh = make(chan struct{})
+	}
+
+	id, err := uuid.New(rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	if opa.logger == nil {
+		opa.logger = logging.NewNoOpLogger()
+	}
+
+	if opa.console == nil {
+		l := logging.NewStandardLogger()
+		l.SetFormatter(&logrus.JSONFormatter{})
+		opa.console = l
+	}
+
+	opa.manager, err = plugins.New(
+		opa.configBytes,
+		id,
+		opa.store,
+		plugins.Logger(opa.logger),
+		plugins.ConsoleLogger(opa.console))
+	if err != nil {
+		return nil, err
+	}
+
+	opa.interQueryBuiltinCache = cache.NewInterQueryCache(opa.manager.InterQueryBuiltinCacheConfig())
+	opa.manager.RegisterPluginStatusListener("sdk", opa.pluginStatusUpdated)
+	opa.manager.RegisterCompilerTrigger(opa.compilerUpdated)
+
+	d, err := discovery.New(opa.manager)
+	if err != nil {
+		return nil, err
+	}
+
+	opa.manager.Register(discovery.Name, d)
+
+	if err := opa.manager.Start(ctx); err != nil {
+		return nil, err
+	}
+
+	if !opa.async {
+		if err := opa.waitForReadiness(ctx); err != nil {
```

Comment:
[nit] I guess we could inline this. It's a bit of a letdown to scroll all the way down to waitForReadiness to find that one select 😉

Reply:
Fixed.